earth_enterprise/src/fusion/autoingest/sysman/gefusion

```diff
@@ -66,9 +66,33 @@ function stop_daemon {
 	else
 	    echo "Failed"
 	fi
+  else
+  echo "Not running"
+  fi
+}
+
+function check_daemon {
+  # more DRY to use "gefdaemoncheck --checkrunning"
+  #   but it seems to assume --checksysman is true
+  #   thus wont work on non-master hosts.
+  pidfile="/var/opt/google/run/$1.pid"
+  if [ -s $pidfile ] ; then
+    pid=`cat $pidfile`
+    ps -ef | grep $pid | grep -q $1
```

Comment:
I think it's safer to test with:

```BASH
if ps --pid "$pid" >/dev/null; then
  returncode=0
else
  returncode=1
fi
```

Reply:
Minor: Also, I think it's shorter to set 1 variable (e.g., `returncode`) when you test a condition, rather than 2 (`returncode` and `message`).

E.g.:
```
if [ "$returncode" = 0 ]; then
  echo "$1:   Running (PID $pid)."
  return 0
else
  echo "$1:   Not running."
  return 1
fi
```

Reply:
I usually run a couple checks (grep'ing for both pid and executable name) because pids can be reused.  When I just need to check for a live pid, I usually use [ -f /proc/$pid ], which requires no special permissions (unlike accessing /proc/$pid/*).  But of course I defer.

That does indeed shorten the code, thanks.

Reply:
Checking for the executable name seems like a nice idea.  You can filter `ps` results by command with the `-C` option.  (E.g., `ps -C bash`.)  However, you can easily cheat, if you're only checking the file name of the executed command.

Probably the really robust approach would be to obtain an exclusive write lock on a PID file in each daemon process, which would be released by the kernel when the process dies, and checking for the lock to tell if a daemon is running.    However, checking the absolute path of a running process should also work pretty well.  (I'm not sure how you could cheat that.)

I don't know how to check the absolute path to the executable with `ps`.  However, on Linux systems, you can use `readlink "/proc/$pid/exe"`.

So, maybe, you can use something like:

```BASH
# $1: Full path to the daemon executable.
function check_daemon {
  # . . .
  daemon_name=$(basename "$1")
  if ps --pid "$pid" >/dev/null && [ "$(readlink "/proc/$pid/exe")" = "$1" ]; then
  # . . .
  echo "$daemon_name:   Running (PID $pid)."
  # . . .
}
```

Reply:
I hadnt thought of security so much as a stale .pid file in an environment of high process turnover, so maybe the old gehttpd pid is now some unrelated process, like a grep.  But either way, I like the added specificity.  As long as we're getting into /proc, we may as well abandon ps:
if [ -f "/proc/$pid" ] && [ "$(readlink "/proc/$pid/exe")" = "$1" ]; then ...
Ah, but we dont have the full path to the daemon executable in the service init script--gestartdaemon knows that.  But I guess it's ok to repeat ourselves since this issue isnt about DRY!  Not sure I have time for another round of build/install/test tonight tho.

Reply:
Yeah, I think abandoning `ps` is fine, since this init script doesn't support anything besides Linux anyway.

Actually, I don't see `gestartdaemon.cpp` implementing any path resolution.  It just relies on the daemon being in the `PATH`.

I.e., if you want to be more robust, you could, e.g.:

```BASH
# $1: Daemon executable name, or path.
function check_daemon {
  # . . .
  daemon_path=$(which "$1")
  deamon_name=$(basename "$daemon_path")
  # . . .
  if [ -d "/proc/$pid" ] && [ "$(readlink "/proc/$pid/exe")" = "$daemon_path" ]; then
    # If you want to be fancy, show the process state:
    daemon_state=$(grep -Po -e '^State:\s*\K.*' "/proc/$pid/status")
    echo "$daemon_name: $daemon_state, PID $pid."
    return 0
  fi
  # . . .
}
```

I'm not a fan of re-implementing Debian's `start-stop-daemon`, or Red Hat's `daemon` commands, but that's another discussion.

Reply:
Thanks, I stand corrected about gestartdaemon.  Isnt daemon_name just "$1"?  And wont daemon state generally be sleeping, at least for the geserver processes that spawn workers?  I suggest that idea be considered out of scope here, maybe as a verbose option on status.

Reply:
Maybe your check of /proc/$pid/status could be the germ of an admin dashboard (if there isnt one already).

Reply:
In order to test my edits, I run:
sudo /etc/init.d/gefusion stop
sudo /etc/init.d/geserver stop
( cd ~/work/$opengee/earthenterprise/earth_enterprise && scons -j2 internal=1 stage_install )
( cd $opengee/earthenterprise/earth_enterprise/src/installer/ && sudo ./install_server.sh && echo $divider && sudo ./install_fusion.sh )
Maybe there's a shorter way, but if not, the install scripts might benefit from the equivalent of a "-y" option (I think I type 'n', 'c', 'c', 'c') so that this sequence can occur unattended.